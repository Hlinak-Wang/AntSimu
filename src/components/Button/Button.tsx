import React from 'react';
import classnames from 'classnames';

export enum ButtonSize {
  small = 'sm',
  large = 'lg',
}

export enum ButtonType {
  primary = 'primary',
  default = 'default',
  link = 'link',
}

interface baseButtonProps {
  className ?: string;
  disabled ?: boolean;
  size?: ButtonSize;
  btnType?: ButtonType;
  children: React.ReactNode;
  href?: string;
  danger?:boolean
}

const Button: React.FC<baseButtonProps> = (props) => {
  const { 
    btnType,
    disabled,
    size,
    children,
    href,
    danger
  } = props;
  const classes = classnames('btn', {
    [`btn-${btnType}`]: btnType,
    [`btn-${size}`]: size,
    [`btn-danger`]: danger,
    'disabled': (btnType === ButtonType.link) && disabled
  })
  if (btnType === ButtonType.link) {
    return (
      <a
        className={classes}
        href = {href}
      >
        {children}
      </a>
    );
  } else {
    return (
      <button
        className={classes}
        disabled = {disabled}
      >
        {children}
      </button>
    )
  }
}

Button.defaultProps = {
  disabled: false,
  danger: false,
  btnType: ButtonType.default
}

export default Button;