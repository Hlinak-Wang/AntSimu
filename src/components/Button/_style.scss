@keyframes outline-default {
  from {
    box-shadow: 0 0 0px 0px transparent;
  }
  15% {
    box-shadow: 0 0 5px 1px darken($primary, 5%);
  }
  to {
    box-shadow: 0 0 10px 0px transparent;
  }
}

@keyframes outline-default-danger {
  from {
    box-shadow: 0 0 0px 0px transparent;
  }
  15% {
    box-shadow: 0 0 5px 1px darken($danger, 5%);
  }
  to {
    box-shadow: 0 0 10px 0px transparent;
  }
}

@mixin btn-style(
  $background,
  $border,
  $fontColor,
  $borderStyle,
  $activeBack: $background,
  $activeBorder: $border,
  $activeColor: $fontColor
) {
  color: $fontColor;
  border-style: $borderStyle;
  border-width: 1px;
  border-color: $border;
  background: $background;

  &:hover {
    background-color: lighten($activeBack, 5%);
    border-color: lighten($activeBorder, 5%);
    color: lighten($activeColor, 5%);
  }
  &:active {
    background-color: darken($activeBack, 5%);
    border-color: darken($activeBorder, 5%);
    color: darken($activeColor, 5%);
  }
}

@mixin button-size($padding, $font-size) {
  font-size: $font-size;
  > a {
    padding: $padding;
  }
}

.btn-default-style {
  @include btn-style($primary, $primary, white, solid);
  &::after {
    content: "";
    position: absolute;
    top: 0; bottom: 0;
    left: 0; right: 0;
    background: transparent;
    box-shadow: 0 0 10px 5px transparent;
  }
  &:focus::after {
    animation: outline-default 0.3s;
  }
  &.btn-danger {
    @include btn-style($danger, $danger, white, solid);
    &:focus::after {
      animation: outline-default-danger 0.3s;
    }
  }
}

.btn-dash-style {
  @include btn-style(transparent, #8c8c8c, #000, dashed, transparent, $primary, $primary);
  &::after {
    content: "";
    position: absolute;
    top: 0; bottom: 0;
    left: 0; right: 0;
    background: transparent;
    box-shadow: 0 0 10px 5px transparent;
  }
  &:focus::after {
    animation: outline-default 0.3s;
  }
  &.btn-danger {
     @include btn-style(transparent, $danger, $danger, dashed);
     &:focus::after {
      animation: outline-default-danger 0.3s;
    }
  }
}

.btn-vacuum-style {
  @include btn-style(transparent, transparent, #000000, hidden, rgba(0,0,0,0.05), transparent);
  
  &.btn-danger {
    @include btn-style(transparent, transparent, #000000, hidden, lighten($danger,20%), transparent, white);
  }
}



.btn {
  display: inline-block;
  position: relative;
  text-align: center;
  white-space: nowrap;
  box-sizing: border-box;
  background-image: none;
  font-family: $btn-font-family;
  border-radius: $btn-border-radius;
  padding: 0;

  cursor: pointer;
  &:focus {
    outline: none;
  }
  &[disabled] {
    cursor: not-allowed;
    opacity: $btn-disabled-opacity;
    box-shadow: none;
    background-color: #f5f5f5;
    border-color:  #d9d9d9;
    border-style: solid;
    color: rgba(0,0,0,.25);

    > * {// prevent all child's event
      pointer-events: none;
    }

    &:hover {
      background-color: #f5f5f5;
      border-color:  #d9d9d9;
      color: rgba(0,0,0,.25);
    }
  }
  > a {
    font-size: inherit;
    line-height: unset;
    display:inline-block;
    color: inherit;
    text-decoration: none;
  }
}

.btn-block {
  width: 100%;
}

.btn-sm-size {
  @include button-size(0 0.5em, 14px)
}

.btn-mid-size {
  @include button-size(0.2em 1em, 16px)
}

.btn-lg-size {
  @include button-size(0.2em 2em, 18px)
}
