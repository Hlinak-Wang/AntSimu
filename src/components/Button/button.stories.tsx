import React from 'react';
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions';
import Button from './button';


storiesOf('button 按钮', module)
  .add('三种基本按钮类型和其不可用状态', () => (
    <>
      <Button type="link">link</Button>
      <Button disabled type="link">link</Button>
      <br />
      <Button type="default">default button</Button>
      <Button disabled type="default">default button</Button>
      <br />
      <Button type="primary">primary button</Button>
      <Button disabled type="primary">primary button</Button>
    </>
  ))
  .add('可设置三种大小', () => {

    return (
      <>
        <input name="size" type="radio" value="sm" />
        <input name="size" type="radio" value="default" checked/>
        <input name="size" type="radio" value="lg"/>
        <Button type="link">link</Button>
        <Button type="default">default button</Button>
        <Button type="primary">primary button</Button>
      </>
    )
  }
    
  )
  .add('危险按钮', () => (
    <>
      <Button danger type="link">link</Button>
      <br />
      <Button danger type="default">default button</Button>
      <br />
      <Button danger type="primary">primary button</Button>
    </>
  ))
  .add('填充整个父级的宽度', () => (
    <> 
      <Button type="link" block>link</Button>
      <Button type="default" block>default button</Button>
      <Button type="primary" block>primary button</Button>
    </>
  ))