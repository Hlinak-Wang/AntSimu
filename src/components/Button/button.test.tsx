import React from 'react';
//import { render, fireEvent } from '@testing-library/react';
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import Button, { ButtonProps } from './button';

const {shallow, mount}=Enzyme
Enzyme.configure({ adapter: new Adapter() })

let event:any;
const disableProps: ButtonProps = {
  disabled: true,
  onClick: jest.fn(),
}

describe('test button component', () => {
  beforeEach(() => {
    event = {
      onClick: jest.fn()
    }
  })
  it('should render default button', () => {
    const wrapper = shallow(<Button {...event}>default</Button>)
    expect(wrapper.children.length).toEqual(1)
    const outlayer = wrapper.find('button')
    const innerlayer = wrapper.find('a')
    expect(outlayer.hasClass(/(?=.*btn)(?=.*btn-default-style)(?=.*btn-mid-size)*./g)).toBeTruthy()
    expect(innerlayer.text()).toEqual('default')

    wrapper.simulate('click')
    expect(event.onClick).toHaveBeenCalled()
  })
  it('should render button for dash style', () => {
    const wrapper = shallow(<Button btnType="dash" className="dash-test" {...event}>dash button</Button>)
    expect(wrapper.children.length).toEqual(1)
    const outlayer = wrapper.find('button')
    expect(wrapper.find('a').text()).toEqual("dash button")
    expect(outlayer.hasClass(/(?=.*btn)(?=.*btn-dash-style)(?=.*dash-test)(?=.*btn-mid-size).*/g)).toBeTruthy()
    expect(wrapper).toMatchSnapshot();

    wrapper.simulate('click')
    expect(event.onClick).toHaveBeenCalled()
  })
  it('should render button for vacuum style', () => {
    const wrapper = shallow(<Button btnType="vacuum" className="vacuum-test another-class" {...event}>vacuum button</Button>)
    expect(wrapper.children.length).toEqual(1)
    const outlayer = wrapper.find('button')
    expect(wrapper.find('a').text()).toEqual("vacuum button")
    expect(outlayer.hasClass(/(?=.*btn)(?=.*btn-vacuum-style)(?=.*vacuum-test)(?=.*btn-mid-size)(?=.*another-class).*/g)).toBeTruthy()
    expect(wrapper).toMatchSnapshot();

    wrapper.simulate('click')
    expect(event.onClick).toHaveBeenCalled()
  })
  it('should render a link when there is a href prop', () => {
    const wrapper = mount(<Button btnType='default' href="url" {...event}>link</Button>)

    expect(wrapper.children.length).toEqual(1)
    const outlayer = wrapper.find('button')
    expect(wrapper.find('a').text()).toEqual("link")
    expect(wrapper.find('a').prop('href')).toEqual('url')
    expect(outlayer.hasClass(/(?=.*btn)(?=.*btn-default-style)(?=.*btn-mid-size).*/g)).toBeTruthy()
    expect(wrapper).toMatchSnapshot();

    wrapper.simulate('click')
    expect(event.onClick).toHaveBeenCalled()
  })
  it('should rendeer a disabled button when disabled is set to true', () => {
    const wrapper = shallow(<Button {...event} disabled>disabled</Button>)

    expect(wrapper.children.length).toEqual(1)
    const outlayer = wrapper.find('button')
    expect(outlayer.prop('disabled')).toBeTruthy()
    expect(wrapper.find('a').text()).toEqual("disabled")
    expect(outlayer.hasClass(/(?=.*btn)(?=.*btn-default-style)(?=.*btn-mid-size).*/g)).toBeTruthy()
    expect(wrapper).toMatchSnapshot();

    outlayer.simulate('click')
    expect(disableProps.onClick).not.toHaveBeenCalled()
    // const element = wrapper.getByText('disabled') as HTMLButtonElement;
    // expect(element).toBeInTheDocument();
    // expect(element.disabled).toBeTruthy();
    // fireEvent.click(element);
    // expect(disableProps.onClick).not.toHaveBeenCalled();
  })
})