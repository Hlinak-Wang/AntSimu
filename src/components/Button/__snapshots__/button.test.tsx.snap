// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test button component should rendeer a disabled button when disabled is set to true 1`] = `ShallowWrapper {}`;

exports[`test button component should render a link when there is a href prop 1`] = `ReactWrapper {}`;

exports[`test button component should render button for dash style 1`] = `ShallowWrapper {}`;

exports[`test button component should render button for vacuum style 1`] = `ShallowWrapper {}`;
